microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceASTValidator.java

```diff
@@ -240,26 +242,41 @@ private void validateRetryAnnotation(NormalAnnotation annotation) throws JavaMod
 			Expression durationUnitExpr = getAnnotationMemberValueExpression(annotation,
 					DURATION_UNIT_RETRY_ANNOTATION_MEMBER);
 			Expression jitterExpr = getAnnotationMemberValueExpression(annotation, JITTER_RETRY_ANNOTATION_MEMBER);
-			Expression jitterDelayUnitExpr = getAnnotationMemberValueExpression(annotation,
+			Expression jitterUnitExpr = getAnnotationMemberValueExpression(annotation,
 					JITTER_DELAY_UNIT_RETRY_ANNOTATION_MEMBER);
 
 			int delayNum = (int) delayExpr.resolveConstantExpressionValue();
 			int maxDurationNum = (int) maxDurationExpr.resolveConstantExpressionValue();
 			int jitterNum = jitterExpr != null ? (int) jitterExpr.resolveConstantExpressionValue() : 0;
 
-			String delayUnit = delayUnitExpr != null ? delayUnitExpr.toString().split("\\.")[1] : null;
-			String maxDurationUnit = durationUnitExpr != null ? durationUnitExpr.toString().split("\\.")[1] : null;
-			String jitterDelayUnit = jitterDelayUnitExpr != null ? jitterDelayUnitExpr.toString().split("\\.")[1]
-					: null;
-
-			Duration delayValue = delayUnitExpr != null ? Duration.of(delayNum, ChronoUnit.valueOf(delayUnit))
+			String delayUnit = null;
+			if (delayUnitExpr != null) {
+				SimpleName delayUnitName = delayUnitExpr instanceof SimpleName ? (SimpleName) delayUnitExpr
+						: ((QualifiedName) delayUnitExpr).getName();
+				delayUnit = delayUnitName.getIdentifier();
+			}
+			Duration delayValue = delayUnit != null ? Duration.of(delayNum, ChronoUnit.valueOf(delayUnit))
 					: Duration.of(delayNum, ChronoUnit.MILLIS);
-			Duration maxDurationValue = durationUnitExpr != null
-					? Duration.of(maxDurationNum, ChronoUnit.valueOf(maxDurationUnit))
+
+			String durationUnit = null;
+			if (durationUnitExpr != null) {
+				SimpleName durationUnitName = durationUnitExpr instanceof SimpleName ? (SimpleName) durationUnitExpr
+						: ((QualifiedName) durationUnitExpr).getName();
+				durationUnit = durationUnitName.getIdentifier();
+			}
+			Duration maxDurationValue = durationUnit != null
+					? Duration.of(maxDurationNum, ChronoUnit.valueOf(durationUnit))
 					: Duration.of(maxDurationNum, ChronoUnit.MILLIS);
-			Duration jitterValue = jitterDelayUnitExpr != null
-					? Duration.of(jitterNum, ChronoUnit.valueOf(jitterDelayUnit))
+
+			String jitterUnit = null;
+			if (durationUnitExpr != null) {
+				SimpleName jitterUnitName = jitterUnitExpr instanceof SimpleName ? (SimpleName) jitterUnitExpr
+						: ((QualifiedName) jitterUnitExpr).getName();
+				jitterUnit = jitterUnitName.getIdentifier();
+			}
+			Duration jitterValue = jitterUnit != null ? Duration.of(jitterNum, ChronoUnit.valueOf(jitterUnit))
 					: Duration.of(jitterNum, ChronoUnit.MILLIS);
+

```

Comment:
https://github.com/eclipse/lsp4mp/pull/191#discussion_r720494385 .

The effective delay is a range, so you'll end up also having a minDelayValue when subtracting the jitterValue from the delayValue.

Reply:
I believe the minDelayValue would have any effect on the error, correct? According to mp-fault-tolerance, the minimum effective delay is bound by 0.

Reply:
Bound by 0 when the jitter would force it to be less than 0 (since such a value would not make sense). For example, if the delay were 2 and the jitter were 5, the range would be `[2-5, 2+5]=[-3,7]`. However, because of that condition, the range becomes [0,7].

Reply:
Are there any diagnostics that should be thrown for this?

Reply:
Would you be fine with just converting the diagnostic to a warning and simply being something like :

"The effective delay may exceed the maximum duration."

Reply:
Whoops, I'm being silly. You're right, this is not an issue. If delay + jitter < maxDuration we don't have to worry about the ranges overlapping.